include/rocksdb/utilities/agg_merge.h

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <deque>
+#include <string>
+#include <vector>
+
+#include "rocksdb/merge_operator.h"
+#include "rocksdb/slice.h"
+
+namespace ROCKSDB_NAMESPACE {
+// A class used to aggregate data per key/value. The plug-in function is
+// implemented and registered using AddAggregator(). And then use it
+// with merge operator created using CreateAggMergeOperator().
+class Aggregator {
+ public:
+  virtual ~Aggregator() {}
+  // The input list is ordered with insertion order, with the latest one
+  // to be oldest. The oldest one might be from Get().
+  // Return whether aggregation succeeded. False for aggregation error.
+  virtual bool Aggregate(const std::vector<Slice>&,
+                         std::string* result) const = 0;
+
+  // True if a partial aggregation should be invoked. Some aggregators
+  // might opt to skip partial aggregation if possible.
+  virtual bool DoPartialAggregate() const { return true; }
+};
+
+// The function adds aggregation plugin by function name. It is used
+// by all the aggregation operator created using CreateAggMergeOperator().
+// It's currently not thread safe to run concurrently with the aggregation
+// merge operator. It is recommended that all the aggregation function
+// is added before calling CreateAggMergeOperator().
+Status AddAggregator(const std::string& function_name,
+                     std::unique_ptr<Aggregator>&& agg);
+
+// Get a merge operator for aggregation.
+// Users can push values to be updated with a merge operand encoded with
+// registered function name and payload using EncodeAggFuncAndValue(),
+// and the merge operator will invoke the aggregation function.
+// An example:
+//
+//    // Assume class ExampleSumAggregator is implemented to do simple sum.
+//    AddAggregator("sum", std::make_unique<ExampleSumAggregator>());
+//    std::shared_ptr<MergeOperator> mp_guard = CreateAggMergeOperator();
+//    options.merge_operator = mp_guard.get();
+//    ...... // Creating DB
+//    db->Put(WriteOptions(), "foo", "100");
+//    db->Merge(WriteOptions(), "foo", EncodeAggFuncAndValue("sum", "200"));
+//    db->Merge(WriteOptions(), "foo", EncodeAggFuncAndValue("sum", "300"));
```

Comment:
Looking at the code, I wonder if it would not be better to have methods such as:
std::string Encode(const std::string& v)
std::string Encode(int64_t v)

Off the Aggregator, rather than this function.  Are you expecting things like IntegerAggregators (sum, etc) to correctly handle strings that are not properly integers?  Perhaps having a basic IntegerAggregator could go a long way to guaranteeing well-formatted data not causing crashes.